README.md
moov-io/auth
===

[![GoDoc](https://godoc.org/github.com/moov-io/auth?status.svg)](https://godoc.org/github.com/moov-io/auth)
[![Build Status](https://travis-ci.com/moov-io/auth.svg?branch=master)](https://travis-ci.com/moov-io/auth)
[![Coverage Status](https://coveralls.io/repos/github/moov-io/auth/badge.svg?branch=master)](https://coveralls.io/github/moov-io/auth?branch=master)
[![Go Report Card](https://goreportcard.com/badge/github.com/moov-io/auth)](https://goreportcard.com/report/github.com/moov-io/auth)
[![Apache 2 licensed](https://img.shields.io/badge/license-Apache2-blue.svg)](https://raw.githubusercontent.com/moov-io/auth/master/LICENSE)

*project is under active development and is not production ready*

This repository holds the authentication service for [moov.io](https://github.com/moov-io). If you find a problem (security or otherwise), please contact us at [`security@moov.io`](mailto:security@moov.io).

The auth project supports various auth methods:
- REST authentication and user sign-up
- OAuth2 exchange (linked to an authenticated user)

Docs: [docs.moov.io](https://docs.moov.io/en/latest/) | [api docs](https://editor.swagger.io/?url=https://raw.githubusercontent.com/moov-io/auth/master/openapi.yaml)

## Project Status

This project is currently pre-production and could change without much notice, however we are looking for community feedback so please try out our code or give us feedback!

## Getting Started / Install

You can download [our docker image `moov/auth`](https://hub.docker.com/r/moov/auth/) from Docker Hub or use this repository. No configuration is required to serve on `localhost:8080`.

Metrics are served at `localhost:9090/metrics` in Prometheus format.

### Configuration

The follow are environment variables can be configured:

**Required**
- `DOMAIN`: Domain to set on cookies.

**Optional**
- `OAUTH2_CLIENTS_DB_PATH`: File path to our oauth2 clients database.
- `OAUTH2_TOKENS_DB_PATH`: File path to our oauth2 tokens database.
- `SQLITE_DB_PATH`: File path to our sqlite database
- `TLS_CERT` and `TLS_KEY`: File paths to TLS certificate and keyfile (in PEM encoding)

### Endpoints

| Method | Path | Description |
|---|---|---|
| GET | /ping | Always returns "PONG". Useful for readness check |
| POST | /users/create | Create a new user. (Signup) |
| GET | /users/login | Verify if a Cookie is valid for a user. |
| POST | /users/login | Login with an email and password.  |
| DELETE | /users/login | Invalidat a user's active cookies. |
| GET | /oauth2/authorize | Verify a Bearer OAuth2 token. |
| [GET&]POST | /oauth2/token | Create a new OAuth2 token. |
| POST | /oauth2/token/create | Create a new OAuth2 client credential set. |

### metrics

| Name | Help Text |
|---|---|
| auth_successes | Count of successful authorizations |
| auth_failures | Count of failed authorizations |
| auth_inactivations | Count of inactivated auths (i.e. user logout) |
| http_errors | Count of how many 5xx errors we send out |
| oauth2_client_generations | Count of auth tokens created |
| oauth2_token_generations | Count of auth tokens created |
| sqlite_connections | How many sqlite connections and what status they're in. |

## Getting Help

 channel | info
 ------- | -------
 [Project Documentation](https://docs.moov.io/en/latest/) | Our project documentation available online.
 Google Group [moov-users](https://groups.google.com/forum/#!forum/moov-users)| The Moov users Google group is for contributors other people contributing to the Moov project. You can join them without a google account by sending an email to [moov-users+subscribe@googlegroups.com](mailto:moov-users+subscribe@googlegroups.com). After receiving the join-request message, you can simply reply to that to confirm the subscription.
Twitter [@moov_io](https://twitter.com/moov_io)	| You can follow Moov.IO's Twitter feed to get updates on our project(s). You can also tweet us questions or just share blogs or stories.
[GitHub Issue](https://github.com/moov-io) | If you are able to reproduce an problem please open a GitHub Issue under the specific project that caused the error.
[moov-io slack](http://moov-io.slack.com/) | Join our slack channel to have an interactive discussion about the development of the project. [Request an invite to the slack channel](https://join.slack.com/t/moov-io/shared_invite/enQtNDE5NzIwNTYxODEwLTRkYTcyZDI5ZTlkZWRjMzlhMWVhMGZlOTZiOTk4MmM3MmRhZDY4OTJiMDVjOTE2MGEyNWYzYzY1MGMyMThiZjg)

## Supported and Tested Platforms

- 64-bit Linux (Ubuntu, Debian), macOS, and Windows

## Contributing

TODO(adam): contrib and CoC docs
Yes please! Please review our [Contributing guide](CONTRIBUTING.md) and [Code of Conduct](CODE_OF_CONDUCT.md) to get started!

Note: This project uses Go Modules, which requires Go 1.11 or higher, but we ship the vendor directory in our repository.

## License

Apache License 2.0 See [LICENSE](LICENSE) for details.


CHANGELOG.md
## v0.2.0 (Released 2018-09-26)

IMPROVEMENTS

- Added `/ping` route.

BUG FIXES

- OAuth2 routes should have been prefixed as `/oauth2/`.
- admin: fix pprof profiles (not all work)

## v0.1.0 (Unreleased)

INITIAL RELEASE

- HTTP Server with oauth, user/pass signup (and auth)
- Prometheus metrics and pprof setup


signup_test.go
package main

import (
	"testing"
)

func TestSignup__email(t *testing.T) {
	cases := []struct {
		input string
		valid bool
	}{
		{"", false},
		{"test@moov.io", true},
	}
	for i := range cases {
		err := validateEmail(cases[i].input)
		if cases[i].valid && err == nil {
			continue // valid
		}
		if !cases[i].valid && err != nil {
			continue // known bad
		}
		t.Errorf("input=%q, err=%v", cases[i].input, err)
	}
}

func TestSignup__pass(t *testing.T) {
	cases := []struct {
		input string
		valid bool
	}{
		{"", false},
		{"superlongpassword", true},
	}
	for i := range cases {
		err := validatePassword(cases[i].input)
		if cases[i].valid && err == nil {
			continue // valid
		}
		if !cases[i].valid && err != nil {
			continue // known bad
		}
		t.Errorf("input=%q, err=%v", cases[i].input, err)
	}
}

func TestSignup__phone(t *testing.T) {
	cases := []struct {
		input string
		valid bool
	}{
		{"", false},
		{"abcdefgh", false},
		{"a0a0a0a0", false},
		{"1a1a1a1a", false},
		{"0123456789", false},
		{"1", false},
		{"10", true},
		{"109", true},
		{"10909", true},
		{"109090", true},
		{"2090999", true},
		{"30909999", true},
		{"409099999", true},
		{"5090999999", true},
		{"60909999999", true},
		{"709099999999", true},
		{"8090999999999", true},
		{"90909999999999", true},
		{"1009099999999999", false},
		{"+14155552671000", true},
		{"+1415555267100001", false},
		{"+1-6174443000", true},
		{"+33 1 5669 6201", true},
		{"49-8994006308", true},
		{"+972-732858700", true},
		{"+81-90-1234-5678", true},
		{"666.666.6666", true},
	}
	for i := range cases {
		err := validatePhone(cases[i].input)
		if cases[i].valid && err == nil {
			continue // valid
		}
		if !cases[i].valid && err != nil {
			continue // known bad
		}
		t.Errorf("input=%q, err=%v", cases[i].input, err)
	}
}


admin/pprof_test.go
package admin

import (
	"fmt"
	"os"
	"testing"
)

func TestAdmin__profileEnabled(t *testing.T) {
	cases := map[string]bool{
		// enable
		"yes":    true,
		" true ": true,
		"":       true,
		// disable
		"no":       false,
		"jsadlsaj": false,
	}
	for value, enabled := range cases {
		os.Setenv("PPROF_TESTING_VALUE", fmt.Sprintf("%v", enabled))
		if v := profileEnabled("TESTING_VALUE"); v != enabled {
			t.Errorf("value=%q, got=%v, expected=%v", value, v, enabled)
		}
	}
}


http.go
// Copyright 2018 The ACH Authors
// Use of this source code is governed by an Apache License
// license that can be found in the LICENSE file.

package main

import (
	"encoding/json"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"time"

	"github.com/gorilla/mux"
)

const (
	// maxReadBytes is the number of bytes to read
	// from a request body. It's intended to be used
	// with an io.LimitReader
	maxReadBytes = 1 * 1024 * 1024

	cookieName = "moov_auth"
	cookieTTL  = 30 * 24 * time.Hour // days * hours/day * hours
)

var (
	// Domain is the domain to publish cookies under.
	// If empty "localhost" is used.
	//
	// The path is always set to /.
	Domain string = os.Getenv("DOMAIN")
)

func init() {
	if Domain == "" {
		Domain = "localhost"
	}
}

// read consumes an io.Reader (wrapping with io.LimitReader)
// and returns either the resulting bytes or a non-nil error.
func read(r io.Reader) ([]byte, error) {
	r = io.LimitReader(r, maxReadBytes)
	return ioutil.ReadAll(r)
}

// encodeError JSON encodes the supplied error
//
// The HTTP status of "400 Bad Request" is written to the
// response.
func encodeError(w http.ResponseWriter, err error) {
	if err == nil {
		return
	}
	w.WriteHeader(http.StatusBadRequest)
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"error": err.Error(),
	})
}

func internalError(w http.ResponseWriter, err error, component string) {
	internalServerErrors.Add(1)
	logger.Log(component, err)
	w.WriteHeader(http.StatusInternalServerError)
}

// extractCookie attempts to pull out our cookie from the incoming request.
// We use the contents to find the associated userId.
func extractCookie(r *http.Request) *http.Cookie {
	if r == nil {
		return nil
	}
	cs := r.Cookies()
	for i := range cs {
		if cs[i].Name == cookieName {
			return cs[i]
		}
	}
	return nil
}

// createCookie generates a new cookie and associates it with the provided
// userId.
func createCookie(userId string, auth authable) (*http.Cookie, error) {
	cookie := &http.Cookie{
		Domain:   Domain,
		Expires:  time.Now().Add(cookieTTL),
		HttpOnly: true,
		Name:     cookieName,
		Path:     "/",
		Secure:   serveViaTLS,
		Value:    generateID(),
	}
	if err := auth.writeCookie(userId, cookie); err != nil {
		return nil, err
	}
	return cookie, nil
}

func addPingRoute(r *mux.Router) {
	r.Methods("GET").Path("/ping").HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain")
		w.Write([]byte("PONG"))
	})
}


user.go
// Copyright 2018 The ACH Authors
// Use of this source code is governed by an Apache License
// license that can be found in the LICENSE file.

package main

import (
	"crypto/rand"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"errors"
	"fmt"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/go-kit/kit/log"
	"golang.org/x/crypto/bcrypt"
)

// TODO(adam): validate email (by sending with approval_code) [probably diff file]

// TODO(adam): having a row in user_approval_codes means user isn't verified. Delete on email approval.

type User struct {
	ID         string    `json:"id"`
	Email      string    `json:"email"`
	FirstName  string    `json:"firstName"`
	LastName   string    `json:"lastName"`
	Phone      string    `json:"phone"`
	CompanyURL string    `json:"companyUrl"`
	CreatedAt  time.Time `json:"createdAt"`
}

var (
	dropPlusExtender = regexp.MustCompile(`(\+.*)$`)
	dropPeriods      = strings.NewReplacer(".", "")
)

const (
	bcryptCostFactor = 10

	// from 'go doc time Time.String', format used in sqlite columns
	serializedTimestampFormat = "2006-01-02 15:04:05.999999999 -0700 MST"
)

func (u *User) cleanEmail() string {
	return cleanEmail(u.Email)
}

// cleanEmail strips all the funky characters from an email address.
//
// Essentially this boils down to the following pattern (ignoring case):
//   [a-z0-9]@[a-z0-9].[a-z]
//
// Callers should be aware of when an empty string is returned.
func cleanEmail(email string) string {
	// split at '@'
	parts := strings.Split(email, "@")
	if len(parts) != 2 {
		return ""
	}

	parts[0] = dropPlusExtender.ReplaceAllString(parts[0], "")
	parts[0] = dropPeriods.Replace(parts[0])

	return strings.ToLower(strings.Join(parts, "@"))
}

// generateID creates a new ID for our auth system.
// Do no assume anything about these ID's other than
// they are strings. Case matters
func generateID() string {
	bs := make([]byte, 20)
	n, err := rand.Read(bs)
	if err != nil || n == 0 {
		logger.Log("generateID", fmt.Sprintf("n=%d, err=%v", n, err))
		return ""
	}
	return strings.ToLower(hex.EncodeToString(bs))
}

type userRepository interface {
	lookupByUserId(id string) (*User, error)

	// lookupByEmail finds a user by the given email address.
	// It's recommended you provide User.cleanEmail()
	lookupByEmail(email string) (*User, error)

	upsert(*User) error
}

type sqliteUserRepository struct {
	db  *sql.DB
	log log.Logger
}

func (s *sqliteUserRepository) lookupByUserId(userId string) (*User, error) {
	query := `select u.email, u.created_at, ud.first_name, ud.last_name, ud.phone, ud.company_url
from users as u
inner join user_details as ud
on u.user_id = ud.user_id
where u.user_id = ?
limit 1`
	stmt, err := s.db.Prepare(query)
	if err != nil {
		return nil, err
	}
	row := stmt.QueryRow(userId)

	u := &User{}
	u.ID = userId
	var createdAt string // needs parsing
	row.Scan(&u.Email, &createdAt, &u.FirstName, &u.LastName, &u.Phone, &u.CompanyURL)
	t, err := time.Parse(serializedTimestampFormat, createdAt)
	if err != nil {
		s.log.Log("user", fmt.Sprintf("bad users.created_at format %q: %v", createdAt, err))
	}
	u.CreatedAt = t
	return u, nil
}

func (s *sqliteUserRepository) lookupByEmail(email string) (*User, error) {
	query := `select user_id from users where clean_email = ?`
	stmt, err := s.db.Prepare(query)
	if err != nil {
		return nil, err
	}
	row := stmt.QueryRow(cleanEmail(email))

	var userId string
	row.Scan(&userId)

	if userId == "" {
		return nil, errors.New("user not found")
	}
	return s.lookupByUserId(userId)
}

func (s *sqliteUserRepository) upsert(inc *User) error {
	tx, err := s.db.Begin()
	if err != nil {
		return err
	}

	// insert/update into 'users'
	query := `replace into users (user_id, email, clean_email, created_at) values (?, ?, ?, ?)`
	stmt, err := tx.Prepare(query)
	if err != nil {
		e := tx.Rollback()
		return fmt.Errorf("problem preparing users query userId=%s, err=%v, rollback err=%v", inc.ID, err, e)
	}
	_, err = stmt.Exec(inc.ID, inc.Email, inc.cleanEmail(), inc.CreatedAt.Format(serializedTimestampFormat))
	if err != nil {
		e := tx.Rollback()
		return fmt.Errorf("problem upserting users userId=%s, err=%v, rollback err=%v", inc.ID, err, e)
	}

	// insert/update into 'user_details'
	query = `replace into user_details (user_id, first_name, last_name, phone, company_url) values (?, ?, ?, ?, ?)`
	stmt, err = tx.Prepare(query)
	if err != nil {
		e := tx.Rollback()
		return fmt.Errorf("problem preparing user_details query userId=%s, err=%v, rollback err=%v", inc.ID, err, e)
	}
	_, err = stmt.Exec(inc.ID, inc.FirstName, inc.LastName, inc.Phone, inc.CompanyURL)
	if err != nil {
		e := tx.Rollback()
		return fmt.Errorf("problem upserting user_details userId=%s, err=%v, rollback err=%v", inc.ID, err, e)
	}

	return tx.Commit()
}

// authable represents the interactions of a user's authentication
// status. This boils down to password comparison and cookie data.
type authable interface {
	findUserId(data string) (string, error)
	invalidateCookies(userId string) error
	writeCookie(userId string, cookie *http.Cookie) error

	// checkPassword compares the provided password for the user.
	// a non-nil error is returned if the passwords don't match
	// or that the userId doesn't exist.
	checkPassword(userId string, pass string) error
	writePassword(userId string, pass string) error
}

type auth struct {
	db  *sql.DB
	log log.Logger
}

// findUserId takes cookie data and returns the userId associated
func (a *auth) findUserId(data string) (string, error) {
	data = strings.TrimSpace(data)
	if data == "" {
		return "", errors.New("no cookie data provided")
	}

	// the SHA256 checksum is stored, not the actual data.
	data, err := hash(data)
	if err != nil {
		return "", err
	}

	query := `select user_id from user_cookies where data == ? and valid_until > ?`
	stmt, err := a.db.Prepare(query)
	if err != nil {
		return "", err
	}
	defer stmt.Close()

	rows, err := stmt.Query(data, time.Now().Format(serializedTimestampFormat))
	if err != nil {
		return "", err
	}
	defer rows.Close()

	var userId string
	for rows.Next() {
		rows.Scan(&userId)
		if userId != "" {
			return userId, nil
		}
	}
	return "", nil
}

func (a *auth) invalidateCookies(userId string) error {
	stmt, err := a.db.Prepare(`delete from user_cookies where user_id = ?`)
	if err != nil {
		return err
	}
	res, err := stmt.Exec(userId)
	if err != nil {
		return err
	}
	n, _ := res.RowsAffected()
	a.log.Log("user", fmt.Sprintf("deleted %d cookies for userId=%s", n, userId))
	return nil
}

func (a *auth) writeCookie(userId string, cookie *http.Cookie) error {
	query := `insert or replace into user_cookies (user_id, data, valid_until) values (?, ?, ?)`
	stmt, err := a.db.Prepare(query)
	if err != nil {
		return err
	}

	// hash the data
	data, err := hash(cookie.Value)
	if err != nil {
		return err
	}
	validUntil := cookie.Expires.Format(serializedTimestampFormat)

	// write row
	stmt.Exec(userId, data, validUntil)
	return nil
}

// fakeBcryptRounds just performs a bcrypt.GenerateFromPassword and then
// a bcrypto.CompareHashAndPassword afterwords. In an attempt to make happy
// and sad paths take "approximately" the same.
func (a *auth) fakeBcryptRounds() {
	id := generateID()
	bcrypt.CompareHashAndPassword([]byte(id), []byte(id))
}

func (a *auth) checkPassword(userId string, incoming string) error {
	stmt, err := a.db.Prepare(`select password, salt from user_passwords where user_id = ?`)
	if err != nil {
		a.fakeBcryptRounds()
		return err
	}

	row := stmt.QueryRow(userId)
	var storedPassword, storedSalt string
	if err := row.Scan(&storedPassword, &storedSalt); err != nil {
		a.fakeBcryptRounds()
		return err
	}

	return bcrypt.CompareHashAndPassword([]byte(storedPassword), []byte(incoming+storedSalt))
}

// writePassword saves a user's password. This function performs no authn/z and generates a new
// salt (which is also saved).
func (a *auth) writePassword(userId string, pass string) error {
	salt := generateID()
	incomingPassword, err := bcrypt.GenerateFromPassword([]byte(pass+salt), bcryptCostFactor)
	if err != nil {
		return err
	}

	stmt, err := a.db.Prepare(`replace into user_passwords (user_id, password, salt) values (?, ?, ?)`)
	if err != nil {
		return err
	}
	_, err = stmt.Exec(userId, string(incomingPassword), salt)
	if err != nil {
		return err
	}
	a.log.Log("user", fmt.Sprintf("userId=%s updated password", userId))
	return nil
}

func hash(in string) (string, error) {
	ss := sha256.New()
	n, err := ss.Write([]byte(in))
	if err != nil {
		return "", err
	}
	if n == 0 {
		return "", err
	}
	return hex.EncodeToString(ss.Sum(nil)), nil
}


login.go
// Copyright 2018 The ACH Authors
// Use of this source code is governed by an Apache License
// license that can be found in the LICENSE file.
package main

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/go-kit/kit/log"
	"github.com/gorilla/mux"
)

type loginRequest struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

func addLoginRoutes(router *mux.Router, logger log.Logger, auth authable, userService userRepository) {
	router.Methods("GET").Path("/users/login").HandlerFunc(checkLogin(logger, auth, userService))
	router.Methods("POST").Path("/users/login").HandlerFunc(loginRoute(logger, auth, userService))
}

func checkLogin(logger log.Logger, auth authable, userService userRepository) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain")

		cookie := extractCookie(r)
		if cookie == nil {
			w.WriteHeader(http.StatusForbidden)
			return
		}
		userId, err := auth.findUserId(cookie.Value)
		if err != nil {
			internalError(w, err, "login")
			return
		}
		if userId == "" {
			w.WriteHeader(http.StatusForbidden)
			return
		}
		w.Header().Set("X-User-Id", userId)
		w.WriteHeader(http.StatusOK)
	}
}

func loginRoute(logger log.Logger, auth authable, userService userRepository) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Body == nil {
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		bs, err := read(r.Body)
		if err != nil {
			internalError(w, err, "login")
			return
		}

		// read request body
		var login loginRequest
		if err := json.Unmarshal(bs, &login); err != nil {
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		// Basic data sanity checks
		if err := validateEmail(login.Email); err != nil {
			encodeError(w, err)
			return
		}
		if err := validatePassword(login.Password); err != nil {
			encodeError(w, err)
			return
		}

		// find user by email
		u, err := userService.lookupByEmail(login.Email)
		if err != nil {
			// Mark this (and password check) as failure only because
			// the user is involved at this point. Otherwise it's their
			// developer's problem (i.e. bad json).
			authFailures.With("method", "web").Add(1)
			w.WriteHeader(http.StatusForbidden)
			return
		}

		// find user by userId and password
		if err := auth.checkPassword(u.ID, login.Password); err != nil {
			authFailures.With("method", "web").Add(1)
			logger.Log("login", fmt.Sprintf("userId=%s failed: %v", u.ID, err))
			w.WriteHeader(http.StatusForbidden)
			return
		}

		// success route, let's finish!
		authSuccesses.With("method", "web").Add(1)
		cookie, err := createCookie(u.ID, auth)
		if err != nil {
			internalError(w, err, "login")
			return
		}
		if cookie == nil {
			logger.Log("login", fmt.Sprintf("nil cookie for userId=%s", u.ID))
			internalError(w, err, "login")
			return
		}
		if err := auth.writeCookie(u.ID, cookie); err != nil {
			internalError(w, err, "login")
			return
		}

		http.SetCookie(w, cookie)
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		if err := json.NewEncoder(w).Encode(u); err != nil {
			internalError(w, err, "login")
			return
		}
	}
}


oauth.go
// Copyright 2018 The ACH Authors
// Use of this source code is governed by an Apache License
// license that can be found in the LICENSE file.

package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"

	"github.com/moov-io/auth/pkg/buntdbclient"

	"github.com/go-kit/kit/log"
	"github.com/gorilla/mux"
	"gopkg.in/oauth2.v3/errors"
	"gopkg.in/oauth2.v3/manage"
	"gopkg.in/oauth2.v3/models"
	"gopkg.in/oauth2.v3/server"
	"gopkg.in/oauth2.v3/store"
)

type oauth struct {
	manager     *manage.Manager
	clientStore *buntdbclient.ClientStore
	server      *server.Server

	logger log.Logger
}

func setupOauthServer(logger log.Logger) (*oauth, error) {
	out := &oauth{
		logger: logger,
	}

	// oauth2 setup
	path := os.Getenv("OAUTH2_TOKENS_DB_PATH")
	if path == "" {
		path = "oauth2_tokens.db"
	}
	tokenStore, err := store.NewFileTokenStore(path)
	if err != nil {
		return nil, fmt.Errorf("problem creating token store: %v", err)
	}

	out.manager = manage.NewDefaultManager()
	out.manager.MapTokenStorage(tokenStore)

	path = os.Getenv("OAUTH2_CLIENTS_DB_PATH")
	if path == "" {
		path = "oauth2_clients.db"
	}
	cs, err := buntdbclient.New(path)
	if err != nil {
		return nil, fmt.Errorf("problem creating clients store: %v", err)
	}
	out.clientStore = cs
	out.manager.MapClientStorage(out.clientStore)

	out.server = server.NewDefaultServer(out.manager)
	out.server.SetAllowGetAccessRequest(true)
	out.server.SetClientInfoHandler(server.ClientFormHandler)
	out.server.SetInternalErrorHandler(func(err error) (re *errors.Response) {
		logger.Log("internal-error", err.Error())
		return
	})
	out.server.SetResponseErrorHandler(func(re *errors.Response) {
		m := re.Error.Error()
		if m == "server_error" || m == "unsupported_grant_type" {
			return
		}
		logger.Log("response-error", m)
	})

	return out, nil
}

// addOAuthRoutes includes our oauth2 routes on the provided mux.Router
func addOAuthRoutes(r *mux.Router, o *oauth, logger log.Logger, auth authable) {
	r.Methods("GET").Path("/oauth2/authorize").HandlerFunc(o.authorizeHandler)
	r.Methods("POST").Path("/oauth2/token/create").HandlerFunc(o.createTokenHandler(auth))

	// Check token routes
	if o.server.Config.AllowGetAccessRequest {
		// only open up GET if the server config asks for it
		r.Methods("GET").Path("/oauth2/token").HandlerFunc(o.tokenHandler)
	}
	r.Methods("POST").Path("/oauth2/token").HandlerFunc(o.tokenHandler)
}

// authorizeHandler checks the request for appropriate oauth information
// and returns "200 OK" if the token is valid.
func (o *oauth) authorizeHandler(w http.ResponseWriter, r *http.Request) {
	// We aren't using HandleAuthorizeRequest here because that assumes redirect_uri
	// exists on the request. We're just checking for a valid token.
	ti, err := o.server.ValidationBearerToken(r)
	if err != nil {
		authFailures.With("method", "oauth2").Add(1)
		encodeError(w, err)
		return
	}
	if ti.GetClientID() == "" {
		authFailures.With("method", "oauth2").Add(1)
		encodeError(w, fmt.Errorf("missing client_id"))
		return
	}

	// Passed token check, return "200 OK"
	authSuccesses.With("method", "oauth2").Add(1)
	w.Header().Set("Content-Type", "text/plain")
	w.WriteHeader(http.StatusOK)
}

// rememberingWriter is a http.ResponseWriter that knows when its headers
// have been written. This is non-idempotent according to http.ResponseWriter,
// but there's no good way to inspect if that's happened yet.
type rememberingWriter struct {
	http.ResponseWriter
	statusCode int
}

// WriteHeader intecepts the embedded WriteHeader call to record what
// status code was written.
func (r *rememberingWriter) WriteHeader(code int) {
	if r.statusCode == 0 {
		r.statusCode = code
	}
	r.ResponseWriter.WriteHeader(code)
}

// tokenHandler passes off the request down to our oauth2 library to
// generate a token (or return an error).
func (o *oauth) tokenHandler(w http.ResponseWriter, r *http.Request) {
	w = &rememberingWriter{ResponseWriter: w}

	// This block is copied from o.server.HandleTokenRequest
	// We needed to inspect what's going on a bit.
	gt, tgr, verr := o.server.ValidationTokenRequest(r)
	if verr != nil {
		encodeError(w, verr)
		return
	}
	ti, verr := o.server.GetAccessToken(gt, tgr)
	if verr != nil {
		encodeError(w, verr)
		return
	}
	data := o.server.GetTokenData(ti)
	bs, err := json.Marshal(data)
	if err != nil {
		encodeError(w, err)
		return
	}
	// (end of copy)

	// HandleTokenRequest currently returns nil even if the token request
	// failed. That menas we can't clearly know if token generation passed or failed.
	//
	// So we need to find out if an error is written, which we can
	// infer by w.WriteHeader call (a 4xx or 5xx status code).
	if ww, ok := w.(*rememberingWriter); ok && ww.statusCode > 400 { // wrote error
		tokenGenerations.Add(1)
		w.Header().Set("X-User-Id", ti.GetUserID()) // only on non-errors
	}

	// Write our response
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	w.Write(bs)
}

// createTokenHandler will create an oauth token for the authenticated user.
//
// This method extracts the user from the cookies in r.
func (o *oauth) createTokenHandler(auth authable) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		userId, err := auth.findUserId(extractCookie(r).Value)
		if err != nil {
			// user not found, return
			w.WriteHeader(http.StatusForbidden)
			return
		}

		records, err := o.clientStore.GetByUserID(userId)
		if err != nil && !strings.Contains(err.Error(), "not found") {
			internalError(w, err, "oauth")
			return
		}
		if len(records) == 0 { // nothing found, so fake one
			records = append(records, &models.Client{})
		}

		clients := make([]*models.Client, len(records))
		for i := range records {
			err = o.clientStore.DeleteByID(records[i].GetID())
			if err != nil && !strings.Contains(err.Error(), "not found") {
				internalError(w, err, "oauth")
				return
			}

			clients[i] = &models.Client{
				ID:     generateID()[:12],
				Secret: generateID(),
				Domain: Domain,
				UserID: userId,
			}

			// Write client into oauth clients db.
			if err := o.clientStore.Set(clients[i].GetID(), clients[i]); err != nil {
				internalError(w, err, "oauth")
				return
			}
		}

		// metrics
		clientGenerations.Add(1)

		// render back new client info
		type response struct {
			Clients []*models.Client `json:"clients"`
		}
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		if err := json.NewEncoder(w).Encode(&response{clients}); err != nil {
			internalError(w, err, "oauth")
			return
		}
	}
}

func (o *oauth) shutdown() error {
	if o == nil || o.clientStore == nil {
		return nil
	}
	return o.clientStore.Close()
}


main.go
// Copyright 2018 The ACH Authors
// Use of this source code is governed by an Apache License
// license that can be found in the LICENSE file.

package main

import (
	"context"
	"crypto/tls"
	"flag"
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/moov-io/auth/admin"

	"github.com/go-kit/kit/log"
	"github.com/go-kit/kit/metrics/prometheus"
	"github.com/gorilla/mux"
	stdprometheus "github.com/prometheus/client_golang/prometheus"
)

var (
	httpAddr = flag.String("http.addr", ":8080", "HTTP listen address")

	logger log.Logger

	// Configuration
	tlsCertificate, tlsPrivateKey = os.Getenv("TLS_CERT"), os.Getenv("TLS_KEY")
	serveViaTLS                   = tlsCertificate != "" && tlsPrivateKey != ""

	// Metrics
	// TODO(adam): be super fancy and generate README.md table in go:generate
	authSuccesses = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
		Name: "auth_successes",
		Help: "Count of successful authorizations",
	}, []string{"method"})
	authFailures = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
		Name: "auth_failures",
		Help: "Count of failed authorizations",
	}, []string{"method"})
	authInactivations = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
		Name: "auth_inactivations",
		Help: "Count of inactivated auths (i.e. user logout)",
	}, []string{"method"})

	internalServerErrors = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
		Name: "http_errors",
		Help: "Count of how many 5xx errors we send out",
	}, nil)

	clientGenerations = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
		Name: "oauth2_client_generations",
		Help: "Count of auth tokens created",
	}, nil)
	tokenGenerations = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
		Name: "oauth2_token_generations",
		Help: "Count of auth tokens created",
	}, nil)
)

func main() {
	flag.Parse()

	// Setup logging, default to stdout
	logger = log.NewLogfmtLogger(os.Stderr)
	logger = log.With(logger, "ts", log.DefaultTimestampUTC)
	logger = log.With(logger, "caller", log.DefaultCaller)
	logger.Log("startup", fmt.Sprintf("Starting auth server version %s", Version))

	// Listen for application termination.
	errs := make(chan error)
	go func() {
		c := make(chan os.Signal, 1)
		signal.Notify(c, syscall.SIGINT, syscall.SIGTERM)
		errs <- fmt.Errorf("%s", <-c)
	}()

	adminServer := admin.NewServer()
	go func() {
		logger.Log("admin", fmt.Sprintf("listening on %s", adminServer.BindAddr()))
		if err := adminServer.Listen(); err != nil {
			err = fmt.Errorf("problem starting admin http: %v", err)
			logger.Log("admin", err)
			errs <- err
		}
	}()
	defer adminServer.Shutdown()

	// migrate database
	db, err := migrate(logger)
	if err != nil {
		logger.Log("admin", err)
		os.Exit(1)
	}
	defer func() {
		if err := db.Close(); err != nil {
			logger.Log("admin", err)
		}
	}()

	oauth, err := setupOauthServer(logger)
	if err != nil {
		logger.Log("oauth", err)
		errs <- err
	}
	defer func() {
		if err := oauth.shutdown(); err != nil {
			logger.Log("oauth", err)
		}
	}()

	// user services
	authService := &auth{
		db:  db,
		log: logger,
	}
	userService := &sqliteUserRepository{
		db:  db,
		log: logger,
	}

	// api routes
	router := mux.NewRouter()
	addPingRoute(router)
	addOAuthRoutes(router, oauth, logger, authService)
	addLoginRoutes(router, logger, authService, userService)
	addLogoutRoutes(router, logger, authService)
	addSignupRoutes(router, logger, authService, userService)
	// TODO(adam): profile CRU[D] routes

	serve := &http.Server{
		Addr:    *httpAddr,
		Handler: router,
		TLSConfig: &tls.Config{
			InsecureSkipVerify:       false,
			PreferServerCipherSuites: true,
			MinVersion:               tls.VersionTLS12,
		},
		ReadTimeout:  30 * time.Second,
		WriteTimeout: 30 * time.Second,
		IdleTimeout:  60 * time.Second,
	}
	shutdownServer := func() {
		if err := serve.Shutdown(context.TODO()); err != nil {
			logger.Log("shutdown", err)
		}
	}
	defer shutdownServer()

	go func() {
		if serveViaTLS {
			logger.Log("transport", "HTTPS", "addr", *httpAddr)
			if err := serve.ListenAndServeTLS(tlsCertificate, tlsPrivateKey); err != nil {
				logger.Log("main", err)
			}
		} else {
			logger.Log("transport", "HTTP", "addr", *httpAddr)
			if err := serve.ListenAndServe(); err != nil {
				logger.Log("main", err)
			}
		}
	}()

	if err := <-errs; err != nil {
		logger.Log("exit", err)
	}
	os.Exit(0)
}
